File: src/main/java/com/chm/plugin/idea/forestx/startup/ForestCheckTask.java
package com.chm.plugin.idea.forestx.startup;

import com.chm.plugin.idea.forestx.annotation.Annotation;
import com.chm.plugin.idea.forestx.constants.Constants;
import com.chm.plugin.idea.forestx.tw.RightSidebarToolWindow;
import com.google.common.collect.Sets;
import com.intellij.openapi.application.ReadAction;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressIndicatorProvider;
import com.intellij.openapi.progress.Task.Backgroundable;
import com.intellij.openapi.project.Project;
import com.intellij.psi.JavaPsiFacade;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiModifierListOwner;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.search.GlobalSearchScopesCore;
import com.intellij.psi.search.ProjectScope;
import com.intellij.psi.search.SearchScope;
import com.intellij.psi.search.searches.AnnotationTargetsSearch;
import com.intellij.util.Query;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.Set;

/**
 * @author caihongming
 * @version v1.0
 * @since 2022-08-09
 **/
public class ForestCheckTask extends Backgroundable {

    ForestCheckTask(Project project) {
        super(project, "Forest check");
    }

    @Override
    public void run(@NotNull ProgressIndicator indicator) {
        ReadAction.nonBlocking(() ->
                        this.runCollectors(ProgressIndicatorProvider.getGlobalProgressIndicator()))
                .inSmartMode(this.myProject)
                .wrapProgress(indicator)
                .executeSynchronously();
    }

    private void runCollectors(ProgressIndicator indicator) {
        Set<PsiClass> searchedSet = Sets.newHashSet();
        GlobalSearchScope librariesScope = ProjectScope.getLibrariesScope(myProject);
        SearchScope inheritorsScope = GlobalSearchScopesCore.projectProductionScope(this.myProject).union(librariesScope);
        for (Annotation annotation : Annotation.FOREST_METHOD_ANNOTATION) {
            PsiClass annotationClass = JavaPsiFacade.getInstance(this.myProject).findClass(annotation.getQualifiedName(), librariesScope);
            if (annotationClass == null) {
                continue;
            }
            Query<PsiModifierListOwner> annotationTargets = AnnotationTargetsSearch.search(annotationClass, inheritorsScope);
            Collection<PsiModifierListOwner> all = annotationTargets.findAll();
            for (PsiModifierListOwner psiModifierListOwner : all) {
                if (psiModifierListOwner instanceof PsiClass) {
                    searchedSet.add((PsiClass) psiModifierListOwner);
                } else if (psiModifierListOwner instanceof PsiMethod) {
                    searchedSet.add(((PsiMethod) psiModifierListOwner).getContainingClass());
                }
            }

            RightSidebarToolWindow mainForm = Constants.getRightSidebar(this.myProject);
            int i = 0;
            for (PsiClass psiClass : searchedSet) {
                mainForm.processClass(psiClass);
                mainForm.afterProcessClass();
                indicator.setFraction((double) (i++) / (double) searchedSet.size());
            }
        }
    }
}

File: src/main/kotlin/com/chm/plugin/idea/forestx/startup/ForestCheckTask.kt
package com.chm.plugin.idea.forestx.startup

import com.chm.plugin.idea.forestx.tw.RightSidebarToolWindow
import com.chm.plugin.idea.forestx.tw.RightSidebarUpdater
import com.google.common.collect.Maps
import com.intellij.openapi.application.ReadAction
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressIndicatorProvider
import com.intellij.openapi.progress.Task.Backgroundable
import com.intellij.openapi.project.Project

/**
 * @author caihongming
 * @version v1.0
 * @since 2022-08-25
 **/
class ForestCheckTask(project: Project) : Backgroundable(project, "Forest check") {

    companion object {
        val TaskMap: MutableMap<Project, ForestCheckTask> = Maps.newConcurrentMap()

        @JvmStatic
        fun getInstance(project: Project): ForestCheckTask? {
            return TaskMap[project]
        }
    }

    init {
        TaskMap[project] = this
    }

    private var checkFinish = false

    private val updater = RightSidebarUpdater.getInstance(project)


    override fun run(indicator: ProgressIndicator) {
        ReadAction.nonBlocking { runCollectors(ProgressIndicatorProvider.getGlobalProgressIndicator()!!) }
            .inSmartMode(project)
            .wrapProgress(indicator)
            .executeSynchronously()
    }

    private fun runCollectors(indicator: ProgressIndicator) {
        checkFinish = false
        // 处理过的class
        val pendingProcessClassSet = updater.doScanClass()
        // 处理class
        pendingProcessClassSet.forEachIndexed { i, psiClass ->
            // 处理
            updater.updateFromElement(psiClass)
            indicator.isIndeterminate = false
            indicator.fraction = i.toDouble() / pendingProcessClassSet.size.toDouble()
        }
        checkFinish = true
    }

    fun isCheckFinish() = checkFinish
}

